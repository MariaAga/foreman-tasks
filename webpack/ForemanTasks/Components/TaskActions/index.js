import { sprintf } from 'foremanReact/common/I18n';
import API from 'foremanReact/API';
import { addToast } from 'foremanReact/redux/actions/toasts';
import {
  FOREMAN_TASK_DETAILS_TOGGLE_UNLOCK_MODAL,
  FOREMAN_TASK_DETAILS_TOGGLE_FORCE_UNLOCK_MODAL,
  TASKS_RESUME_REQUEST,
  TASKS_RESUME_SUCCESS,
  TASKS_RESUME_FAILURE,
  TASKS_CANCEL_REQUEST,
  TASKS_CANCEL_SUCCESS,
  TASKS_CANCEL_FAILURE,
} from './TaskActionsConstants';
import { TOAST_TYPES } from '../common/ToastTypesConstants';
import {
  resumeToastInfo,
  cancelToastInfo,
  toastDispatch,
} from './TaskActionHelpers';

export const toggleUnlockModal = () => ({
  type: FOREMAN_TASK_DETAILS_TOGGLE_UNLOCK_MODAL,
});
export const toggleForceUnlockModal = () => ({
  type: FOREMAN_TASK_DETAILS_TOGGLE_FORCE_UNLOCK_MODAL,
});

export const cancelTaskRequest = (id, name) => async dispatch => {
  dispatch(
    addToast({
      type: TOAST_TYPES.INFO,
      message: sprintf('Trying to cancel %s task', name),
    })
  );
  dispatch({ type: TASKS_CANCEL_REQUEST });
  try {
    await API.post(`/foreman_tasks/tasks/${id}/cancel`);
    dispatch({ type: TASKS_CANCEL_SUCCESS });
    toastDispatch({
      type: 'cancelled',
      name,
      toastInfo: cancelToastInfo,
      dispatch,
    });
  } catch (error) {
    dispatch({ type: TASKS_CANCEL_FAILURE, payload: error });
    toastDispatch({
      type: 'skipped',
      name,
      toastInfo: cancelToastInfo,
      dispatch,
    });
  }
};

export const resumeTaskRequest = (id, name) => async dispatch => {
  dispatch({ type: TASKS_RESUME_REQUEST });
  try {
    await API.post(`/foreman_tasks/tasks/${id}/resume`);

    dispatch({ type: TASKS_RESUME_SUCCESS });
    toastDispatch({
      type: 'resumed',
      name,
      toastInfo: resumeToastInfo,
      dispatch,
    });
  } catch (error) {
    dispatch({ type: TASKS_RESUME_FAILURE, payload: error });
    toastDispatch({
      type: 'failed',
      name,
      toastInfo: resumeToastInfo,
      dispatch,
    });
  }
};
