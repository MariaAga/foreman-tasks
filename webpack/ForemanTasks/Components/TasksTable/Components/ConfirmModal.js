import React from 'react';
import PropTypes from 'prop-types';
import { startCase } from 'lodash';
import { sprintf, translate as __ } from 'foremanReact/common/I18n';
import { Button } from 'patternfly-react';
import ForemanModal from 'foremanReact/components/ForemanModal';

export const ConfirmModal = ({
  closeModal,
  actionText,
  actionState,
  action,
  selectedRowsLen,
  id,
}) => (
  <ForemanModal
    title={sprintf(__('%s Selected Tasks'), startCase(actionText))}
    id={id}
  >
    {sprintf(
      __(
        `This will %(action)s %(number)s task(s), putting them in the %(state)s state. Are you sure?`
      ),
      { action: actionText, number: selectedRowsLen, state: actionState }
    )}
    <ForemanModal.Footer>
      <Button onClick={closeModal}>{__('No')}</Button>
      <Button
        bsStyle="primary"
        onClick={() => {
          action();
          closeModal();
        }}
      >
        {__('Yes')}
      </Button>
    </ForemanModal.Footer>
  </ForemanModal>
);

ConfirmModal.propTypes = {
  actionText: PropTypes.string.isRequired,
  actionState: PropTypes.string.isRequired,
  closeModal: PropTypes.func.isRequired,
  selectedRowsLen: PropTypes.number.isRequired,
  action: PropTypes.func,
  id: PropTypes.string.isRequired,
};

ConfirmModal.defaultProps = {
  action: () => null,
};

export default ConfirmModal;
