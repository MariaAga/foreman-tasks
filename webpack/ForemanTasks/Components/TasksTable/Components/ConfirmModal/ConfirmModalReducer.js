import Immutable from 'seamless-immutable';
import {
  UPDATE_MODAL,
  CANCEL_SELECTED_MODAL,
  RESUME_SELECTED_MODAL,
  RESUME_MODAL,
  CANCEL_MODAL,
} from '../../TasksTableConstants';

const initialState = Immutable({});

export const ConfirmModalReducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case UPDATE_MODAL:
      switch (payload.modalID) {
        case CANCEL_SELECTED_MODAL:
        case CANCEL_MODAL:
          return state.merge({
            actionText: 'cancel',
            actionState: 'stopped',
            actionType: payload.modalID,
          });
        case RESUME_SELECTED_MODAL:
        case RESUME_MODAL:
          return state.merge({
            actionText: 'resume',
            actionState: 'running',
            actionType: payload.modalID,
          });
        default:
          return state.set('actionType', payload.modalID);
      }
    default:
      return state;
  }
};
export default ConfirmModalReducer;
