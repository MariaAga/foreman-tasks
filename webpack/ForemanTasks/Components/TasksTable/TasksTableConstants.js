import { getControllerSearchProps } from 'foremanReact/constants';

export const TASKS_TABLE_ID = 'TASKS_TABLE';
export const TASKS_CONTROLLER = 'tasks';
export const TASKS_SUCCESS = 'TASKS_SUCCESS';
export const TASKS_REQUEST = 'TASKS_REQUEST';
export const TASKS_FAILURE = 'TASKS_FAILURE';
export const SELECT_ROWS = 'SELECT_ROWS';
export const UNSELECT_ROWS = 'UNSELECT_ROWS';
export const UNSELECT_ALL_ROWS = 'UNSELECT_ALL_ROWS';
export const CANCEL = 'CANCEL';
export const RESUME = 'RESUME';
export const CLOSED = 'CLOSED';

export const TASKS_TABLE_SELECTED_MODAL = 'TASKS_TABLE_SELECTED_MODAL';
export const TASKS_SEARCH_PROPS = getControllerSearchProps('tasks');
