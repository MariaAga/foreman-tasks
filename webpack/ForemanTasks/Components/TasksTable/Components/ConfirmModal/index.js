import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { ConfirmModal } from './ConfirmModal';
import reducer from './ConfirmModalReducer';
import tasksActions from './ConfirmModalActions';
import {
  selectActionText,
  selectActionState,
  selectActionType,
  selectClicked,
  selectSelectedRowsLen,
} from './ConfirmModalSelectors';
import { selectAllRowsSelected } from '../../TasksTableSelectors';

const mapStateToProps = state => ({
  actionText: selectActionText(state),
  actionType: selectActionType(state),
  actionState: selectActionState(state),
  allRowsSelected: selectAllRowsSelected(state),
  clicked: selectClicked(state),
  selectedRowsLen: selectSelectedRowsLen(state),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(tasksActions, dispatch);

export const reducers = { confirmModal: reducer };

export default connect(mapStateToProps, mapDispatchToProps)(ConfirmModal);
