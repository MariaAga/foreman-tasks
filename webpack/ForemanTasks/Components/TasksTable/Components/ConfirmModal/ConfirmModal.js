import React from 'react';
import PropTypes from 'prop-types';
import { startCase } from 'lodash';
import { sprintf, translate as __ } from 'foremanReact/common/I18n';
import { Button } from 'patternfly-react';
import ForemanModal from 'foremanReact/components/ForemanModal';
import { FORCE_UNLOCK_MODAL } from '../../../TaskActions/TaskActionsConstants';
import { FORCE_UNLOCK_SELECTED_MODAL } from '../../TasksTableConstants';
import { ForceUnlockModal } from '../../../TaskActions/UnlockModals';

export const ConfirmModal = ({
  actionText,
  actionState,
  actionType,
  selectedRowsLen,
  id,
  parentTaskID,
  url,
  uriQuery: query,
  setModalClosed,
  ...props
}) => {
  if ([FORCE_UNLOCK_MODAL, FORCE_UNLOCK_SELECTED_MODAL].includes(actionType)) {
    return (
      <ForceUnlockModal
        onClick={() => {
          props[actionType]({ url, query, parentTaskID });
          setModalClosed();
        }}
        id={id}
        selectedRowsLen={selectedRowsLen}
      />
    );
  }
  return (
    <ForemanModal
      title={sprintf(__('%s Selected Tasks'), startCase(actionText))}
      id={id}
    >
      {sprintf(
        __(
          `This will %(action)s %(number)s task(s), putting them in the %(state)s state. Are you sure?`
        ),
        { action: actionText, number: selectedRowsLen, state: actionState }
      )}
      <ForemanModal.Footer>
        <Button onClick={setModalClosed}>{__('No')}</Button>
        <Button
          className="confirm-button"
          bsStyle="primary"
          onClick={() => {
            props[actionType]({ url, query, parentTaskID });
            setModalClosed();
          }}
        >
          {__('Yes')}
        </Button>
      </ForemanModal.Footer>
    </ForemanModal>
  );
};

ConfirmModal.propTypes = {
  actionText: PropTypes.string,
  actionState: PropTypes.string,
  selectedRowsLen: PropTypes.number.isRequired,
  actionType: PropTypes.string,
  id: PropTypes.string.isRequired,
  parentTaskID: PropTypes.string,
  url: PropTypes.string.isRequired,
  uriQuery: PropTypes.object,
  setModalClosed: PropTypes.func.isRequired,
};

ConfirmModal.defaultProps = {
  actionType: '',
  actionText: '',
  actionState: '',
  parentTaskID: '',
  uriQuery: {},
};

export default ConfirmModal;
