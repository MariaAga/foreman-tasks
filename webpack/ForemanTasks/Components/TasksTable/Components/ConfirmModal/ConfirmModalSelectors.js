import { selectForemanTasks } from '../../../../ForemanTasksSelectors';
import {
  selectTasksTableQuery,
  selectResults,
  selectSelectedRows,
  selectItemCount,
  selectAllRowsSelected,
} from '../../TasksTableSelectors';
import { RESUME_MODAL, CANCEL_MODAL } from '../../TasksTableConstants';

export const selectCofirmModal = state =>
  selectForemanTasks(state).confirmModal || {};

export const selectActionType = state => selectCofirmModal(state).actionType;
export const selectActionText = state => selectCofirmModal(state).actionText;
export const selectActionState = state => selectCofirmModal(state).actionState;
export const selectClicked = state =>
  selectTasksTableQuery(state).clicked || {};

export const selectSelectedTasks = state => {
  const selectedIDs = selectResults(state).filter(item =>
    selectSelectedRows(state).includes(item.id)
  );
  return selectedIDs.map(item => ({
    name: item.action,
    id: item.id,
    isCancellable: item.availableActions.cancellable,
    isResumable: item.availableActions.resumable,
  }));
};

export const selectSelectedRowsLen = state => {
  if ([CANCEL_MODAL, RESUME_MODAL].includes(selectActionType(state))) {
    return 1;
  }
  if (selectAllRowsSelected(state)) {
    return selectItemCount(state);
  }
  return selectSelectedRows(state).length;
};
