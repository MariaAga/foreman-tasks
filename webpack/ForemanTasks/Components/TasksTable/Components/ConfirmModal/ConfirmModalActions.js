import {
  resumeTask,
  cancelTask,
  forceCancelTask,
} from '../../TasksTableActions';
import {
  bulkCancelBySearch,
  bulkCancelById,
  bulkResumeBySearch,
  bulkResumeById,
  bulkForceCancelBySearch,
  bulkForceCancelById,
} from '../../TasksBulkActions';
import { selectClicked, selectSelectedTasks } from './ConfirmModalSelectors';
import { selectAllRowsSelected } from '../../TasksTableSelectors';
import {
  CANCEL_MODAL,
  RESUME_MODAL,
  CANCEL_SELECTED_MODAL,
  RESUME_SELECTED_MODAL,
  FORCE_UNLOCK_SELECTED_MODAL,
} from '../../TasksTableConstants';
import { FORCE_UNLOCK_MODAL } from '../../../TaskActions/TaskActionsConstants';

export default {
  [CANCEL_SELECTED_MODAL]: ({ url, query, parentTaskID }) => (
    dispatch,
    getState
  ) => {
    if (selectAllRowsSelected(getState())) {
      return dispatch(bulkCancelBySearch({ query, parentTaskID }));
    }
    return dispatch(
      bulkCancelById({
        selected: selectSelectedTasks(getState()),
        url,
        parentTaskID,
      })
    );
  },
  [CANCEL_MODAL]: ({ url, parentTaskID }) => (dispatch, getState) => {
    const { taskId, taskName } = selectClicked(getState());
    return dispatch(
      cancelTask({
        taskId,
        taskName,
        url,
        parentTaskID,
      })
    );
  },
  [RESUME_SELECTED_MODAL]: ({ url, query, parentTaskID }) => (
    dispatch,
    getState
  ) => {
    if (selectAllRowsSelected(getState())) {
      return dispatch(bulkResumeBySearch({ query, parentTaskID }));
    }
    return dispatch(
      bulkResumeById({
        selected: selectSelectedTasks(getState()),
        url,
        parentTaskID,
      })
    );
  },

  [RESUME_MODAL]: ({ url, parentTaskID }) => (dispatch, getState) => {
    const { taskId, taskName } = selectClicked(getState());
    return dispatch(
      resumeTask({
        taskId,
        taskName,
        url,
        parentTaskID,
      })
    );
  },

  [FORCE_UNLOCK_MODAL]: ({ url, parentTaskID }) => (dispatch, getState) => {
    const { taskId, taskName } = selectClicked(getState());
    return dispatch(
      forceCancelTask({
        taskId,
        taskName,
        url,
        parentTaskID,
      })
    );
  },
  [FORCE_UNLOCK_SELECTED_MODAL]: ({ url, query, parentTaskID }) => (
    dispatch,
    getState
  ) => {
    if (selectAllRowsSelected(getState())) {
      return dispatch(bulkForceCancelBySearch({ query, parentTaskID }));
    }
    return dispatch(
      bulkForceCancelById({
        selected: selectSelectedTasks(getState()),
        url,
        parentTaskID,
      })
    );
  },
};
