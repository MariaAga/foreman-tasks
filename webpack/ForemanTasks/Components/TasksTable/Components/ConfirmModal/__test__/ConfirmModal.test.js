import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import { ConfirmModal } from '../ConfirmModal';
import { RESUME_SELECTED_MODAL } from '../../../TasksTableConstants';
import { FORCE_UNLOCK_MODAL } from '../../../../TaskActions/TaskActionsConstants';

const fixtures = {
  'renders ConfirmModal': {
    actionType: RESUME_SELECTED_MODAL,
    actionText: 'some text',
    actionState: 'some state',
    selectedRowsLen: 1,
    id: 'modalID',
    parentTaskID: 'parent-id',
    allRowsSelected: true,
    url: 'some-url',
    uriQuery: { state: 'stopped' },
    setModalClosed: jest.fn(),
  },

  'renders ConfirmModal for unlock ': {
    actionType: FORCE_UNLOCK_MODAL,
    actionText: 'some text',
    actionState: 'some state',
    selectedRowsLen: 1,
    id: 'modalID',
    parentTaskID: 'parent-id',
    allRowsSelected: true,
    url: 'some-url',
    uriQuery: { state: 'stopped' },
    setModalClosed: jest.fn(),
  },
};

describe('ConfirmModal', () => {
  testComponentSnapshotsWithFixtures(ConfirmModal, fixtures);
});
