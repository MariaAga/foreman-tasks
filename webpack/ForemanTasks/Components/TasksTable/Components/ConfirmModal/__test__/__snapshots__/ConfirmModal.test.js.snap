// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmModal renders ConfirmModal 1`] = `
<ForemanModal
  id="modalID"
  title="Some Text Selected Tasks"
>
  This will some text 1 task(s), putting them in the some state state. Are you sure?
  <Component>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      onClick={[MockFunction]}
    >
      No
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      className="confirm-button"
      disabled={false}
      onClick={[Function]}
    >
      Yes
    </Button>
  </Component>
</ForemanModal>
`;

exports[`ConfirmModal renders ConfirmModal for unlock  1`] = `
<ForceUnlockModal
  id="modalID"
  onClick={[Function]}
  selectedRowsLen={1}
/>
`;
